Add rich text control

lukecarbis opened the pull request:
Resolves #20.

Commit:
Add rich text control

RobStino commented:
@lukecarbis See above for conflicts notice.

RobStino commented:
@lukecarbis Without this option checked, I still get new paragraphs when I create a new line.
![image](https://user-images.githubusercontent.com/1578148/51501885-69d69200-1e1f-11e9-80cb-fbf62fdc6772.png)
![image](https://user-images.githubusercontent.com/1578148/51501900-7955db00-1e1f-11e9-895d-b2960180ddfc.png)

RobStino commented:
I noticed that the Rich Text controls disappear when typing. This felt weird, but then I checked the normal paragraph block and noticed that the behaviour is the same there.
See this vid for what I mean
https://d.pr/v/Ja6liq

Also, is it possible to get the text alignment options?
![image](https://user-images.githubusercontent.com/1578148/51502061-0ac54d00-1e20-11e9-8475-043ef306d6f3.png)

Commit:
Merge branch 'develop' into feature/rich-text

RobStino commented:
Hey @lukecarbis 
Testing this further this morning and saw this bug.
Basically, I can't seem to go back to an empty state for the rich text field.
https://d.pr/v/0k0AuY

kienstra commented:
**Looks Good**
*Alignment Buttons*

Hi @lukecarbis,
This PR looks good. If it would help, I'm working on a commit that adds alignment buttons:

<img width="979" alt="alignment-buttons" src="https://user-images.githubusercontent.com/4063887/54088277-6f822b80-432a-11e9-9ce9-01f3488bfab5.png">

Commit:
Merge branch 'develop' into feature/rich-text

davidallenlewis commented:
This is awesome. Looking forward to it being merged.

I would like to see a format menu for headers though and a list option. You can use separate blocks for headings and lists but if your rich text is part of a larger "format" style block (say a 2 column image + text block with a background colour / image) it would be nice to have lists and headers. No more than that though. Keep it simple.

dave-smyth commented:
This might be a nightmare to implement, but it would be amazing if there was some way to control the options that are available in the editor. For instance, in some cases it would be useful to limit them to multiple lines of text only (no formatting other than bold and italic, perhaps), other times headings might be useful and in some cases alignment might be good, too.

Perhaps the editor could have three settings (minimal, standard, full) that would enable different levels of control.

Minimal: multiple lines, bold, italic, links
Standard: minimal + headings, lists
Full: everything else

AndrewSheetMetal commented:
Hey guys, just a quick question:

We need this feature for a project. How is the progress? When can we expect this feature to be distributed in a release?
Thanks!

lukecarbis commented:
@AndrewSheetMetal We're hoping to release it in our 1.3 milestone, which we're working on right now. :)

kienstra committed:
Merge in develop, resolve conflicts

There were 2 conflicts:
js/blocks/loader/editor.scss
 - retained both edits
php/post-types/class-block-post.php
 - retained the edits in the develop branch,
as the means of instantiating controls changed.

kienstra committed:
Improve how controls are stored

Use the $name property in each class
to store them in the  property,
instead of the name in $control_names.

kienstra committed:
Add test class for Rich_Text

This is very straight-forward,
and very similar to other control text classes.

kienstra commented:
**Question About Alignment Controls**

Hi @lukecarbis and @RobStino,
Hope you had a great weekend.

Would you like the Rich Text control to have alignment controls?

Below is the current Rich Text control, with the possible location of alignment controls.

I liked the [location here](https://github.com/getblocklab/block-lab/pull/221#issuecomment-471322765) better, maybe the API changed in the meantime.

![rich-text-current](https://user-images.githubusercontent.com/4063887/58399814-ff2ca180-801e-11e9-921b-02939e81bed4.gif)

I'm not sure if it's possible to get the `RichText` to display the alignment selected, though. For example, on clicking "Align text right," it'd be best if the `RichText` showed everything aligned to the right:

<img width="928" alt="alignment-here" src="https://user-images.githubusercontent.com/4063887/58400005-87ab4200-801f-11e9-851f-fafaae389014.png">

If you'd like alignment controls, I'll look at this more.

Thanks!

kienstra commented:
This PR already worked pretty well, even before I made some commits above. The [.gif above](https://github.com/getblocklab/block-lab/pull/221#issuecomment-496096523) shows the current state.

All of the controls shown above are included in the `RichText` by default, including Code, Inline Image, and Strikethrough. Except for the alignment controls shown above.

lukecarbis commented:
@kienstra I like the earlier version of the alignment controls better too (where they're not hidden behind a dropdown). But if that's no longer easily possible, I don't have that strong feelings about it.

Showing how the alignment effects the text, however, is very important. Without it, I'm sure I would think it was a bug.

lukecarbis commented on js/blocks/controls/index.js:
```diff
@@ -18,6 +19,7 @@ import BlockLabUserControl from './user'
 export default {
 	text: BlockLabTextControl,
 	textarea: BlockLabTextareaControl,
+	rich_text: BlockLabRichTextControl,
```
Our first two word control!

kienstra replied:
Haha, a big occasion!

lukecarbis commented on php/blocks/controls/class-rich-text.php:
```diff
@@ -0,0 +1,78 @@
+				'sanitize' => 'sanitize_text_field',
+			)
+		);
+		$this->settings[] = new Control_Setting(
```
Hmm… what do you think… should we?

a) Remove this option entirely, and force new paragraphs. This is "rich text" and so paragraphs should be expected
b) Enable this by default. I would expect paragraphs in rich text
c) Use the same setting labels / options as we recently implemented for the textarea (doesn't work quite the same though since the textarea isn't an imported control with a `multiline` option)
d) Leave as it (but maybe better align the Help text and Label to the textarea setting?)

lukecarbis replied:
@RobStino I'd love your thoughts on this, too. I'm leaning toward option A. What's a use case for not automatically creating paragraphs with a rich text field?

lukecarbis commented on php/blocks/controls/class-rich-text.php:
```diff
@@ -0,0 +1,78 @@
+	 * @return void
+	 */
+	public function register_settings() {
+		$this->settings[] = new Control_Setting(
```
We should also find a way of disabling the "Field Location" setting, or forcing the Editor option, since it simply doesn't work when loaded in the Inspector.

![Screen Shot 2019-05-28 at 8 26 46 am](https://user-images.githubusercontent.com/1097667/58441077-59faf180-8122-11e9-9051-d06e01ffed42.png)

kienstra replied:
Ah, good catch.

lukecarbis commented on php/post-types/class-block-post.php:
```diff
@@ -128,7 +133,8 @@ public function get_control( $control_name ) {
 
-		$control_class = 'Block_Lab\\Blocks\\Controls\\' . ucwords( $control_name );
+		$class_name    = ucwords( $control_name, '_' );
+		$control_class = 'Block_Lab\\Blocks\\Controls\\' . $class_name;
```
Nice.

lukecarbis commented on tests/php/blocks/controls/test-class-image.php:
```diff
@@ -22,7 +22,7 @@ class Test_Image extends \WP_UnitTestCase {
 	 * Instance of the setting.
 	 *
-	 * @var Controls\Control_setting
+	 * @var Controls\Control_Setting
```
Great pickup!

kienstra replied:
Haha, thanks!

kienstra commented:
Hi @lukecarbis,
Thanks for looking at this.

>@kienstra I like the earlier version of the alignment controls better too (where they're not hidden behind a dropdown). But if that's no longer easily possible, I don't have that strong feelings about it.

Yeah, that display was better. I'll look at if it's still possible.

>Showing how the alignment effects the text, however, is very important. Without it, I'm sure I would think it was a bug.

Good point. I'll also work on this.

lukecarbis reviewed (commented):
I can't approve / request changes since this is my PR… but so far it looks great. Haven't yet been able to test alignment controls. Is that committed yet? I couldn't find them.

![Screen Shot 2019-05-28 at 8 16 17 am](https://user-images.githubusercontent.com/1097667/58441138-bb22c500-8122-11e9-948b-fba6c20cc2d2.png)

A few minor suggestions, plus getting the alignment controls working is important. Alignment + Paragraphs is the challenging part, I believe!

kienstra commented:
>Haven't yet been able to test alignment controls. Is that committed yet?

No, I just have it locally in a messy commit 😄It's not working in the sense of applying the alignment in the editor.

RobStino commented:
Hey all.
Testing this and it's looking good. Everything so far is behaving as expected.
Re the field setting for having `enter` start a new paragraph. I think this should be the default and actually be removed as a setting altogether. I can't think of a good use case otherwise. Also, `shift` + `enter` creates a break tag anyway so I think we're covered.
Looking forward to testing the alignment controls.

RobStino commented:
Seeing that the control just doesn't work in the Inspector, I think we don't make it an option for this particular field. For now at least.

kienstra commented:
>Re the field setting for having enter start a new paragraph. I think this should be the default and actually be removed as a setting altogether. I can't think of a good use case otherwise. Also, shift + enter creates a break tag anyway so I think we're covered.

Good point. Unless there are other ideas, I'll remove the 'Use Paragraphs' option for now.

kienstra committed:
Remove the 'Use Paragraphs' checkbox

As Rob and Luke mentioned,
this isn't very helpful for the Rich Text control.

kienstra committed:
Add the location in the control class, not by default

kienstra committed:
Add a 'location' setting to almost every control, make this a little more DRY

Set a $settings_config with the common
config values, like 'help', 'location', and 'placeholder'.
This will make each of the controls a little more DRY.
But there will be a need to test this thoroughly.
@todo: add PHPUnit tests for the rest of the controls.

kienstra committed:
Add alignment controls, though they're not working yet

These now appear alongside the bold, italic,
and link controls.
But their value isn't saving yet,
and it doesn't have an effect on the RichText yet.

kienstra commented:
**Alignment Controls Are In PR**
*Though They're Not Working Yet*

With https://github.com/getblocklab/block-lab/pull/221/commits/881d85706e4efd79bf69ee53ff6b5795241eeca7, there are now alignment controls, though I still have to get them to work 😄 I'll let you know how it goes.

<img width="849" alt="alignment-in-pr" src="https://user-images.githubusercontent.com/4063887/58462202-032bf280-80f7-11e9-931c-8834efd0dcd1.png">

kienstra committed:
Add left, center, and right alignment controls to Rich Text

There are still issues to consider:
* Deselect any other alignment control when one
is selected
* When aligning right, nothing can be to the left.

kienstra committed:
Set multiline to true, instead of using a field value

There's no field value for this,
so simply set this to true.

kienstra commented:
**WIP Alignment Controls**

The latest commits added alignment controls. But I still need to do some refactoring and bugfixes.

![align-controls-here](https://user-images.githubusercontent.com/4063887/58472263-7e979f00-810b-11e9-8ca3-8be7cefffb09.gif)

kienstra committed:
Use the simpler AlignmentToolbar for the alignment controls

This will be cleaner than the previous approach,
which was essentially reimplementing them.
Alignment is mainly working well,
though there are some bugs.

kienstra committed:
Add JSDoc descriptions and tags, rename constansts

Use ALL_UPPERCASE for constants,
and make the JSDocs more descriptive.

kienstra committed:
Allow de-selecting an alignment control

Now, you can "un-click" an alignment control,
in addition to disabling it by clicking
a different alignment control.

kienstra commented:
**Current State**
*Alignment Controls Work*

![rich-text-here](https://user-images.githubusercontent.com/4063887/58537671-f1128880-81b8-11e9-8177-9762f5ff18fe.gif)

kienstra commented on js/blocks/controls/rich-text.js:
```diff
@@ -0,0 +1,120 @@
+			{
+			/*
+			* @todo: Resolve known issue with toolbar not disappearing on blur
+			* @see: https://github.com/WordPress/gutenberg/issues/7463
```
It looks like there was some progress with this in https://github.com/WordPress/gutenberg/pull/14640, but there might not be a clear path to fixing this yet.

kienstra commented on php/blocks/controls/class-checkbox.php:
```diff
@@ -44,15 +44,8 @@ public function __construct() {
-				'sanitize' => 'sanitize_text_field',
-			)
-		);
+		$this->settings[] = new Control_Setting( $this->settings_config['location'] );
```
@lukecarbis, your idea of adding the location setting only when needed was good. This does that.

kienstra commented on php/blocks/controls/class-color.php:
```diff
@@ -37,23 +37,8 @@ public function __construct() {
-		);
+		foreach ( array( 'location', 'help', 'default' ) as $setting ) {
+			$this->settings[] = new Control_Setting( $this->settings_config[ $setting ] );
+		}
```
This is intended to only refactor these controls, not change them. In brief testing of a few of them, there wasn't a backwards-compatibility problem.

But it'd be good if I added unit tests for the remaining controls before the next release. Maybe this will be in a different PR, for simplicity.

kienstra committed:
Minor changes to DocBlocks

Change a method description,
and add an underscore to the class.

kienstra committed:
Add PHPUnit tests for 2 new methods

render_settings_location()
and sanitize_location()

kienstra committed:
Display the location again in the .block-fields-row

Before, I had hidden this.
So show it again.

lukecarbis committed:
Merge branch 'develop' into feature/rich-text

lukecarbis commented:
@kienstra I've got a bug that's blocking me from testing. When editing a block, if I choose any field type, no options load.

![2019-05-29 18-35-37 2019-05-29 18_36_08](https://user-images.githubusercontent.com/1097667/58542331-c53ae580-8240-11e9-84f5-b43b982835e4.gif)

kienstra commented:
Hi @lukecarbis,
Sorry about that. https://github.com/getblocklab/block-lab/pull/221/commits/50361cadcc659329f437f707667ba6dea5e931a5 should fix the issue with the settings not loading.

Ah, just a minute, I didn't push the commit yet.

kienstra committed:
Fix an issue where many of the settings didn't appear

Simply change a selector in the JS file
that handles the AJAX request.

kienstra commented:
@lukecarbis, https://github.com/getblocklab/block-lab/pull/221/commits/5bf3c02c4088c0722785aab00e14fca781443bbd should fix the issue.

kienstra commented:
Hi @lukecarbis,
Thanks for reviewing this. 

As [shown above](https://github.com/getblocklab/block-lab/pull/221#issuecomment-496818207), there might still be some styling work needed here.

The `<ServerSideRender>` of the Rich Text control doesn't look exactly like the `<RichText>`. I'm not sure how close it's possible to get, but there could definitely be work remaining on that.

lukecarbis commented:
Cool – thanks! Got another bug now. The block edit state isn't showing anything. `npm run build` runs without any problems, but doesn't help.

![2019-05-29 18-49-59 2019-05-29 18_50_31](https://user-images.githubusercontent.com/1097667/58543406-c66d1200-8242-11e9-82c3-4e52cf7840ca.gif)

kienstra commented:
OK, I see that also. I'm working on it.

kienstra committed:
A quick workaround to enable testing to continue, must be reverted

As Luke mentioned, the control isn't appearing
So this is a hack to allow it to appear.

kienstra commented:
Hi @lukecarbis, https://github.com/getblocklab/block-lab/pull/221/commits/ba38e356ae8075b27397a3491d6904706b8227a0 should allow you to continue testing. Of course, I'll have to revert that and come up with a better solution.

kienstra committed:
Improvement on previous workaround

This still proabbly needs more consideration,
like maybe setting a default value for the location.

RobStino commented:
Just did some testing. Overall, great work! This is looking sharp!
Couple of things I noticed that may need some attention.

- The alignment action requires you to select the whole line. I believe normal editor behaviour doesn't need this. It just requires the cursor to be within the paragraph you are aligning.
- Pressing enter adds a `</br` tag rather than a new paragraph as we have done in the textarea
- Some styling improvements on the buttons.

Walkthrough if you need examples. Sorry about the _hushed_ voice. Kids just got to bed. :D
https://d.pr/v/qv2w8p

kienstra committed:
Move Rich Text to Pro controls

kienstra committed:
For Rich Text control, use wpautop() for the front-end output

As Rob suggested,
the previous <br> tags weren't consistent
with <p> tags in the Text Area.

kienstra commented:
Hi @RobStino,
Thanks a lot for testing this.

>Pressing enter adds a </br tag rather than a new paragraph as we have done in the textarea

Good catch. With https://github.com/getblocklab/block-lab/pull/221/commits/135afefe5f5f7fb033c86acc86b1c866ded55acf, this uses `wpautop()` on the front-end output to create `<p>` tags.

>The alignment action requires you to select the whole line. I believe normal editor behaviour doesn't need this. It just requires the cursor to be within the paragraph you are aligning.

>Some styling improvements on the buttons.

Good points. I'll work on them.

kienstra commented:
Hi @RobStino,

>Some styling improvements on the buttons.

Could you please elaborate a little on that? There definitely could be some improvements needed 😄

lukecarbis commented:
+1 for not selecting the whole line of text to realign it.

Buttons looks good to me. They match the paragraph block. Just one very minor glitch:
<img width="270" alt="Screen Shot 2019-05-30 at 7 53 10 am" src="https://user-images.githubusercontent.com/1097667/58594113-23eb7800-82b0-11e9-8969-0702afecbc44.png">

I'm seeing some screwy behaviour with new lines after alignment. Are we intending to break aligned text into a div? Should it be a paragraph?

![2019-05-30 07-49-56 2019-05-30 07_54_25](https://user-images.githubusercontent.com/1097667/58594139-31a0fd80-82b0-11e9-9c0f-ab8b7a4f50e6.gif)

Also, is it possible to style the text in the rich text area to better match a standard textarea? Something as simple as reducing the font size would makes a big difference.

<img width="781" alt="Screen Shot 2019-05-30 at 8 00 02 am" src="https://user-images.githubusercontent.com/1097667/58594447-fce17600-82b0-11e9-83f8-e5f286ae170c.png">

lukecarbis commented on js/blocks/controls/rich-text.js:
```diff
@@ -0,0 +1,120 @@
+										{
+											type: ALIGNMENT_CONTROL_NAME,
+											attributes: {
+												align: getAlignmentStyle( newAlignment ),
```
Heavy nesting… but I'm okay with it.

lukecarbis replied:
You could potentially spin out the `onChange` handler into a separate method. But no real need.

kienstra replied:
Yeah, good point.

kienstra replied:
https://github.com/getblocklab/block-lab/pull/221/commits/4afb3f90c1d420195ed2980c77d20671ed812306#diff-b64eb5a514d6a8f70b18938222cd7a4fR45 applies your suggestion of creating a separate method for this.

lukecarbis commented on js/blocks/controls/rich-text.js:
```diff
@@ -0,0 +1,120 @@
+				defaultValue={field.default}
+				value={attr[ field.name ]}
+				className='input-control'
+				multiline={false}
```
If we change this to true, what effect does that have? Would a keyboard "return" create a new paragraph (instead of a line break)?

kienstra replied:
Good point, https://github.com/getblocklab/block-lab/pull/221/commits/f7983f6ae0bf97d362a197153cc0fedd27e9791f changes it to `true`. That helped a lot, as the `RichText` now uses `<p>` for line breaks, just like the front-end.

lukecarbis commented on php/blocks/controls/class-email.php:
```diff
@@ -37,15 +37,8 @@ public function __construct() {
-				'sanitize' => 'sanitize_text_field',
-			)
-		);
+		$this->settings[] = new Control_Setting( $this->settings_config['location'] );
```
Nice! ❤️

kienstra replied:
Thanks!

lukecarbis commented:
Just watched @RobStino's video, and I just want to agree with one thing he was trying to say: it should be possible to _toggle_ alignment, with the alignment controls.

So if I click "left align",   the text should explicitly align left. If I click "left align" again, the text should loose its specified alignment.

kienstra committed:
Enable aligning without selecting text

As Rob mentioned, this is standard for
other UIs, like the Classic Editor
and Paragraph block.
Also, ensure that you can toggle alignment on
and off

kienstra committed:
Fix an off-by-one error in alignment

When aligning text near the end of the Rich Text control,
it often doesn't select the last character.

kienstra committed:
Make the RichText markup more like the front-end markup

As Luke mentioned, the multiline
property of RichText seems to ensure that is
uses <p> for newlines.
Also, make some changes for this to work.
Like look for a different character in the text
for the newline.

kienstra commented:
Hi @lukecarbis,
>Just watched @RobStino's video, and I just want to agree with one thing he was trying to say: it should be possible to toggle alignment, with the alignment controls.

>So if I click "left align",   the text should explicitly align left. If I click "left align" again, the text should loose its specified alignment.

Sure, toggling alignment should work now. Also, it's no longer necessary to select text, simply placing a cursor on a line and clicking the 'center' button should center-align it.

kienstra committed:
Reduce the font size in the Rich Text control to match the Text Area

As Luke mentioned, the Text Area control
has a font-size of 13px.
So match that in the Rich Text control.

kienstra committed:
Remove the extra border, only keeping the border-right

This keeps the same border-right
that is applied with native Gutenberg styles.

kienstra commented:
Hi @lukecarbis,
Thanks a lot for reviewing this.

>+1 for not selecting the whole line of text to realign it.

Sure, that's applied now. 

>Buttons looks good to me. They match the paragraph block. Just one very minor glitch:

How does the fix in https://github.com/getblocklab/block-lab/pull/221/commits/d3901de844de557d9fd7a28b768dd1cce6138798 look?

<img width="421" alt="border-here" src="https://user-images.githubusercontent.com/4063887/58614920-4fa23a00-827f-11e9-98cb-5742e8765075.png">

>I'm seeing some screwy behaviour with new lines after alignment. Are we intending to break aligned text into a div? Should it be a paragraph?

Good point, https://github.com/getblocklab/block-lab/pull/221/commits/f7983f6ae0bf97d362a197153cc0fedd27e9791f makes aligned text and new lines `<p>` tags.

>Also, is it possible to style the text in the rich text area to better match a standard textarea? Something as simple as reducing the font size would makes a big difference.

Yeah, `13px` would make this look more like the Text Area control. 

To me, it looks a little small, though:

<img width="788" alt="13-px-font-size" src="https://user-images.githubusercontent.com/4063887/58615274-46fe3380-8280-11e9-87f4-d8ba5ad79f7b.png">

lukecarbis commented:
@RobStino Can we get a third opinion on the text size? I prefer it smaller (matching a textarea).

lukecarbis commented:
Looks great! Some thoughts here: https://www.fromsmash.com/n-8KJbnS-f-c0

RobStino commented:
Hey friends
Epic work Ryan btw!
Couple of things:
- I agree with 13px. Keeping font size consistent between similar style fields is important I think. 
- the grey highlight you get is the same experience you get when formatting text in a regular text Block. I’m guessing this is where this has come from. 
- this carry over behaviour is probably what is causing the issue Luke spotted when he added a character to the same line as the centre aligned text.

kienstra committed:
In the RichText control, allow selecting formats from their border

As Luke pointed out,
an alignment format should be selected even
if the cursor is at the border,
like the end of a line.
It should be possible to put the cursor at the
end of a centered line and enter more centered text.
This implements that.

kienstra committed:
When the cursor is at the end of the line, engage the formatting

If the line is centered, and the cursor is
at the end of it,
ensure the center formatting control
is active, and can be toggled off.

kienstra commented:
The commits above partially address the testing comments. There's still more work needed, though.

kienstra commented:
Finishing the alignment controls (left, center, and right) could take at least a few more days. The behavior for alignment controls is a little different from the other controls, like bold. 

For example, as mentioned above, placing a cursor at the end of a centered line and typing line should result in more centered text, not a new line.

But the built-in behavior of the `RichText`, like for bold or italic, is that it doesn't continue the formatting. It wouldn't be bold on placing a cursor at the end of bold text and typing. So I've [been overriding](https://github.com/getblocklab/block-lab/pull/221/commits/c630cd125bed025cd76c49e8ae924113c3536341#diff-6a451811c6282091d849a83572f58a36R16) the `RichText` component to accommodate the alignment controls.

kienstra committed:
Add the FetchInput control, which I forgot to stage before

Before, I simply removed this,
where I should have moved it.

lukecarbis commented:
Just had a look through the code. Obviously having to extend the rich text is not ideal, but your implementation is really neat, there's obviously no other way to get alignment working the way we'd like it to.

Nice work!

kienstra commented:
>Obviously having to extend the rich text is not ideal, but your implementation is really neat, there's obviously no other way to get alignment working the way we'd like it to.

Thanks, @lukecarbis! Didn't want it to seem like I was complaining, just explaining it was taking a while 😄

kienstra committed:
Make changes to class extending  RichText

kienstra commented:
**Extending RichText**

Extending the [RichText component](https://github.com/WordPress/gutenberg/blob/788a3d2ea7d5490f149bf5ebe191cbbcb140c6f1/packages/block-editor/src/components/rich-text/index.js#L105) isn't working locally so far.

Extending it looks to be necessary for [certain behaviors](https://github.com/getblocklab/block-lab/pull/221#issuecomment-498601148) of the alignment controls to work.

For example, I've been working on a class:

```js
const { RichText } = wp.editor;
const { LEFT, RIGHT } = wp.keycodes;
const { getComputedStyle, isCollapsed } = wp.richText;

/**
 * Browser dependencies
 */
const { getSelection } = window;

class BlockLabRichText extends RichText {


	/**
	 * Constructs the component class.
	 */
	constructor() {
		super( ...arguments );
		this.onSelectionChange = this.onSelectionChange.bind( this );
		this.onKeyDown = this.onKeyDown.bind( this );
		this.handleHorizontalNavigation = this.handleHorizontalNavigation.bind( this );
		this.onPointerDown = this.onPointerDown.bind( this );
	}

	/**
	 * Overrides the parent implementation of the `selectionchange` handler.
	 *
	 * Mainly copies the parent implemetation.
	 * The main difference is that this allows selecting the entire format by clicking on its border.
	 */
	onSelectionChange() {
		...
	}

	/* etc... */
}
```

But this class isn't using `onSelectionChange` above or the other extending methods. It's simply using the [onSelectionChange()](https://github.com/WordPress/gutenberg/blob/788a3d2ea7d5490f149bf5ebe191cbbcb140c6f1/packages/block-editor/src/components/rich-text/index.js#L458) method in the parent [RichText](https://github.com/WordPress/gutenberg/blob/788a3d2ea7d5490f149bf5ebe191cbbcb140c6f1/packages/block-editor/src/components/rich-text/index.js#L105) class when I debug this.

I'll keep looking at this, but it might not be possible to override these methods in the class to get the expected behavior for the alignment controls.

lukecarbis commented:
@kienstra It might be best to pull alignment controls entirely in that case, and put them in a separate issue.

kienstra committed:
Remove the alignment controls

As we've discussed,
this needs different behavior than the
default alignment controls,
like bold and italic.

kienstra commented:
**Alignment Controls Removed**

Hi @lukecarbis,
Thanks a lot for your ideas for the alignment controls.

I couldn't get them to work, including with the `onChangeAlign()` method.

This would probably need to extend [RichText](https://github.com/WordPress/gutenberg/blob/788a3d2ea7d5490f149bf5ebe191cbbcb140c6f1/packages/block-editor/src/components/rich-text/index.js#L105) somehow, but like we talked about, that's not recommended in React.

The behavior of the default formats like bold is different from the alignment formats in a few ways. I don't see a way to accommodate the alignment formats.

For example:
![centered-this](https://user-images.githubusercontent.com/4063887/59312591-506d9f80-8c73-11e9-9cc6-c842515a013c.gif)

lukecarbis commented:
Without alignment controls, this works brilliantly. I can't approve, since I started the request, but this has my green tick.

Just one topic of discussion. What do you think about reducing the gap between paragraphs from `0.1em` to something a little lower? `0.67em` looks a little nicer, in my opinion, and still provides a clear distinction between paragraph break and line break. Very minor difference:

 
![Screen Shot 2019-06-12 at 9 33 29 am](https://user-images.githubusercontent.com/1097667/59313752-8c494900-8cf5-11e9-98ae-f1b319397868.png)
![Screen Shot 2019-06-12 at 9 33 06 am](https://user-images.githubusercontent.com/1097667/59313754-8e130c80-8cf5-11e9-8fa4-6880d6a24b6b.png)

kienstra committed:
Decrease p margin-top and bottom, Applying Luke's suggestion

This looks better,
as there isn't quite so much space in the Rich Text
control between the paragraphs (line-breaks).

kienstra commented:
Hi @lukecarbis,
Good idea. How does https://github.com/getblocklab/block-lab/pull/221/commits/a5856e53156952ffe8913363ba82d8ebc259fbac look? 

It uses you idea of reducing the margins:

<img width="600" alt="paragraph-tags" src="https://user-images.githubusercontent.com/4063887/59321013-14980180-8c96-11e9-878c-1f49d1157cad.png">

Also, it removes the `FORMATTING_CONTROLS` constant, as it was simply the [default value](https://github.com/WordPress/gutenberg/blob/0cc2e02811b3fc64e5365927d16a105debc29895/packages/block-editor/src/components/rich-text/index.js#L1106) in `RichText`.

lukecarbis commented:
Great work Ryan.

✅ Review passed.

kienstra commented:
Thanks, Luke! Thanks a lot for your work on this.

RobStino reviewed (approved):
Hi everyone.
Sorry for the big delays on reviewing this PR.
Tested in my local and everything looks really good!

*Adding a Rich Text Field to a new Block*
![image](https://user-images.githubusercontent.com/1578148/59645354-b8147500-91b4-11e9-833f-f424f0090f6b.png)

*Adding the block in the Editor*
![image](https://user-images.githubusercontent.com/1578148/59645368-ca8eae80-91b4-11e9-8d4e-9c3530f63689.png)

*Previewing in the Editor*
![image](https://user-images.githubusercontent.com/1578148/59645383-d7130700-91b4-11e9-8416-2e4d01cfab53.png)

*On the front end*
![image](https://user-images.githubusercontent.com/1578148/59645399-e09c6f00-91b4-11e9-98d3-1637109b1544.png)

kienstra commented:
**Thanks!**

Hi @RobStino, 
All good, thanks a lot for testing this!

alancwoo commented:
Is this block type only available in early access/pro? I just installed `1.3.4` and don't see this as an option.

kienstra commented:
Hi @alancwoo, yes, the Rich Text field is only available in the Pro version.
